pyensign/connection.py

```diff
@@ -62,66 +63,71 @@ def __init__(self, connection):
 
         self.stub = ensign_pb2_grpc.EnsignStub(connection.channel)
 
-    # TODO: Publish should not block
-    def publish(self, events):
-        for rep in self.stub.Publish(events):
+    async def publish(self, events):
+        async def next():
+            for event in events:
+                yield event
+
+        async for rep in self.stub.Publish(next()):
             yield rep
 
-    # TODO: Subscribe should not block
-    def subscribe(self, topic_ids, consumer_id="", consumer_group=None):
+    async def subscribe(self, topic_ids, consumer_id="", consumer_group=None):
         open_stream = ensign_pb2.OpenStream(
             topics=topic_ids, consumer_id=consumer_id, group=consumer_group
         )
         subscription = ensign_pb2.Subscription(open_stream=open_stream)
-        for event in self.stub.Subscribe(iter([subscription])):
+
+        # TODO: This is a bidirectional stream, we should be a good citizen and return
+        # acks and nacks to the server.
```

Comment:
:-) absolutely! 

Do we have a story for this? Potentially two stories - we want an "automated" mode where acks are automatically sent back right after the yield; and a "user mode" where the user has to explicitly send acks and nacks back themselves.

Reply:
I created one, thanks!